File: ScriptsToProcess/Auto_Phase.py
"""
AUTO-PHASE ASSIGNER (V5 - STRICT SEMANTIC SEQUENCING)
================================================================================
Reads referenced Rhino curves and assigns independent Phase sequences.
* FIXED: Drops physical Z-proximity checks. Now strictly assigns the podium 
  first (0, 1, 2...), finds the highest podium phase, and starts ALL individual 
  tower stacks sequentially from that exact number.
* MAINTAINS: Persistent logging and UI timing.

INPUTS:
    Guids (Guid) [List Access] : Referenced Rhino geometry IDs.
    Run   (bool) [Item Access] : Wire a Button here to execute.

OUTPUTS:
    Log (str)
================================================================================
"""

import Rhino
import Grasshopper as gh
import System
import time
import scriptcontext as sc

ghenv.Component.Name = "Auto-Phase Assigner by Z"
ghenv.Component.NickName = "Auto_Phase"
ghenv.Component.Description = "Strictly sequences the podium first, then cascades that count to independent towers."

def assign_phases():
    exec_start = time.time()
    
    # Setup unique cache keys for this specific component instance
    uid = str(ghenv.Component.InstanceGuid)
    log_key = "autophase_log_" + uid
    msg_key = "autophase_msg_" + uid
    
    guids = globals().get('Guids')
    run = globals().get('Run')
    
    # Return cached data if the button is not actively being pressed
    if not run:
        cached_log = sc.sticky.get(log_key, ["Awaiting execution. Click the 'Run' button."])
        cached_msg = sc.sticky.get(msg_key, "{}\nAwaiting Run".format(ghenv.Component.NickName))
        return cached_log, cached_msg
        
    if not guids:
        msg = "{}\nNo Guids".format(ghenv.Component.NickName)
        return ["No Guids provided."], msg

    doc = Rhino.RhinoDoc.ActiveDoc
    buildings = {}
    log = []
    
    # 1. Gather and Group the Geometry by BuildingID
    for gid in guids:
        obj = doc.Objects.FindId(gid)
        if not obj: continue
            
        b_id = obj.Attributes.GetUserString("BuildingID") or "Building_01"
        obj_type = obj.Attributes.GetUserString("Type") or "Block"
        
        if obj_type.lower() == "core":
            continue
            
        bbox = obj.Geometry.GetBoundingBox(True)
        z_height = round(bbox.Min.Z, 3)
        t_id = obj.Attributes.GetUserString("TowerID") or "Main"
        
        if b_id not in buildings:
            buildings[b_id] = {'podium_blocks': [], 'tower_blocks': {}}
            
        item = {'obj': obj, 't_id': t_id, 'z': z_height}
        
        # Semantically sort Podium vs Towers
        if "podium" in t_id.lower():
            buildings[b_id]['podium_blocks'].append(item)
        else:
            if t_id not in buildings[b_id]['tower_blocks']:
                buildings[b_id]['tower_blocks'][t_id] = []
            buildings[b_id]['tower_blocks'][t_id].append(item)

    modify_count = 0
    
    # 2. Assign Phases Logically (Podium First, then Towers)
    for b_id, data in buildings.items():
        
        max_podium_phase = -1 # Defaults to -1 so if there is no podium, towers start at 0
        
        # A. Process the Podium First
        podiums = sorted(data['podium_blocks'], key=lambda x: x['z'])
        for i, p_item in enumerate(podiums):
            current_phase = i
            max_podium_phase = current_phase # Store the highest phase reached
            
            p_item['obj'].Attributes.SetUserString("Phase", str(current_phase))
            p_item['obj'].CommitChanges()
            modify_count += 1
            log.append("-> {} ({}): Assigned Phase {}".format(b_id, p_item['t_id'], current_phase))
            
        # B. Process Individual Towers (Cascading from max_podium_phase)
        for t_id, t_items in data['tower_blocks'].items():
            t_items_sorted = sorted(t_items, key=lambda x: x['z'])
            
            for i, t_item in enumerate(t_items_sorted):
                # Start from the podium's max phase + 1, and increment up
                new_phase = max_podium_phase + 1 + i
                new_phase_str = str(new_phase)
                
                t_item['obj'].Attributes.SetUserString("Phase", new_phase_str)
                t_item['obj'].CommitChanges()
                modify_count += 1
                log.append("-> {} (Tower: {}): Assigned Phase {}".format(b_id, t_id, new_phase_str))

    # 3. Format Output Status and UI Message
    if modify_count > 0:
        status_line = "Updated {} phases".format(modify_count)
        log.insert(0, "SUCCESS: " + status_line)
    else:
        status_line = "No valid blocks found"
        log.insert(0, "WARNING: Found no blocks to phase.")
        
    exec_time = (time.time() - exec_start) * 1000
    
    ui_msg = "{}\nTime: {:.0f} ms\n{}".format(
        ghenv.Component.NickName, 
        exec_time, 
        status_line
    )
    
    # Save to persistent memory
    sc.sticky[log_key] = log
    sc.sticky[msg_key] = ui_msg
        
    return log, ui_msg

# Execute
output_log, ui_message = assign_phases()

# Outputs
Log = output_log
ghenv.Component.Message = ui_message

File: Components/AutoPhase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Diagnostics;
using Grasshopper.Kernel;
using Rhino;
using Rhino.DocObjects;

namespace Enzyme.Components
{
    public class AutoPhaseComponent : GH_Component
    {
        private List<string> _cachedLog = new List<string> { "Awaiting execution. Click the 'Run' button." };
        private string _cachedMsg = "Auto_Phase\nAwaiting Run";

        public AutoPhaseComponent()
          : base("Auto-Phase Assigner by Z", "Auto_Phase",
              "Strictly sequences the podium first, then cascades that count to independent towers.",
              "Enzyme", "Masterplan (Beta)")
        {
        }

        public override GH_Exposure Exposure => GH_Exposure.primary;

                public override void AddedToDocument(GH_Document document)
        {
            base.AddedToDocument(document);
            if (this.Attributes == null) this.CreateAttributes();

            bool hasSources = false;
            foreach (var param in this.Params.Input)
                if (param.SourceCount > 0) { hasSources = true; break; }

            if (!hasSources)
            {
                int ix = 220, ox = 250;
                Enzyme.Utils.AutoWireHelper.WireButton(this, document, 1, ix, -150);
            }
        }

        protected override void RegisterInputParams(GH_InputParamManager pManager)
        {
            pManager.AddGenericParameter("Guids", "Guids", "Referenced Rhino geometry IDs.", GH_ParamAccess.list);
            pManager.AddBooleanParameter("Run", "Run", "Wire a Button here to execute.", GH_ParamAccess.item, false);
        }

        protected override void RegisterOutputParams(GH_OutputParamManager pManager)
        {
            pManager.AddTextParameter("Log", "Log", "Execution log", GH_ParamAccess.list);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            List<Grasshopper.Kernel.Types.IGH_Goo> goos = new List<Grasshopper.Kernel.Types.IGH_Goo>();
            if (!DA.GetDataList(0, goos)) return;
            
            bool run = false;
            if (!DA.GetData(1, ref run)) return;

            if (!run)
            {
                DA.SetDataList(0, _cachedLog);
                Message = _cachedMsg;
                return;
            }

            if (goos == null || goos.Count == 0)
            {
                _cachedMsg = $"{this.NickName}\nNo Guids";
                _cachedLog = new List<string> { "No Guids provided." };
                DA.SetDataList(0, _cachedLog);
                Message = _cachedMsg;
                return;
            }

            var doc = RhinoDoc.ActiveDoc;
            if (doc == null) return;

            var buildings = new Dictionary<string, BuildingData>();
            var log = new List<string>();
            int modifyCount = 0;

            Stopwatch sw = Stopwatch.StartNew();

            foreach (var goo in goos)
            {
                Guid gid = Guid.Empty;
                if (goo == null) continue;
                
                if (!goo.CastTo(out gid))
                {
                    if (goo.CastTo(out string strId) && Guid.TryParse(strId, out Guid parsedGuid))
                    {
                        gid = parsedGuid;
                    }
                }
                
                if (gid == Guid.Empty) continue;

                var obj = doc.Objects.FindId(gid);
                if (obj == null) continue;

                string bId = obj.Attributes.GetUserString("BuildingID");
                if (string.IsNullOrEmpty(bId)) bId = "Building_01";

                string objType = obj.Attributes.GetUserString("Type");
                if (string.IsNullOrEmpty(objType)) objType = "Block";

                if (objType.Equals("core", StringComparison.OrdinalIgnoreCase))
                    continue;

                var bbox = obj.Geometry.GetBoundingBox(true);
                double zHeight = Math.Round(bbox.Min.Z, 3);

                string tId = obj.Attributes.GetUserString("TowerID");
                if (string.IsNullOrEmpty(tId)) tId = "Main";

                if (!buildings.ContainsKey(bId))
                    buildings[bId] = new BuildingData();

                var item = new BlockItem { Obj = obj, TowerId = tId, Z = zHeight };

                if (tId.ToLower().Contains("podium"))
                {
                    buildings[bId].PodiumBlocks.Add(item);
                }
                else
                {
                    if (!buildings[bId].TowerBlocks.ContainsKey(tId))
                        buildings[bId].TowerBlocks[tId] = new List<BlockItem>();
                    buildings[bId].TowerBlocks[tId].Add(item);
                }
            }

            foreach (var kvp in buildings)
            {
                string bId = kvp.Key;
                var data = kvp.Value;

                int maxPodiumPhase = -1;

                var podiums = data.PodiumBlocks.OrderBy(x => x.Z).ToList();
                for (int i = 0; i < podiums.Count; i++)
                {
                    int currentPhase = i;
                    maxPodiumPhase = currentPhase;

                    podiums[i].Obj.Attributes.SetUserString("Phase", currentPhase.ToString());
                    podiums[i].Obj.CommitChanges();
                    modifyCount++;
                    log.Add($"-> {bId} ({podiums[i].TowerId}): Assigned Phase {currentPhase}");
                }

                foreach (var tKvp in data.TowerBlocks)
                {
                    string tId = tKvp.Key;
                    var tItems = tKvp.Value.OrderBy(x => x.Z).ToList();

                    for (int i = 0; i < tItems.Count; i++)
                    {
                        int newPhase = maxPodiumPhase + 1 + i;
                        string newPhaseStr = newPhase.ToString();

                        tItems[i].Obj.Attributes.SetUserString("Phase", newPhaseStr);
                        tItems[i].Obj.CommitChanges();
                        modifyCount++;
                        log.Add($"-> {bId} (Tower: {tId}): Assigned Phase {newPhaseStr}");
                    }
                }
            }

            string statusLine;
            if (modifyCount > 0)
            {
                statusLine = $"Updated {modifyCount} phases";
                log.Insert(0, "SUCCESS: " + statusLine);
            }
            else
            {
                statusLine = "No valid blocks found";
                log.Insert(0, "WARNING: Found no blocks to phase.");
            }

            sw.Stop();
            long execTime = sw.ElapsedMilliseconds;

            _cachedMsg = $"{this.NickName}\nTime: {execTime} ms\n{statusLine}";
            _cachedLog = log;

            Message = _cachedMsg;
            DA.SetDataList(0, _cachedLog);
        }

        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                return IconLoader.Load("Auto_Phase.png");
            }
        }

        public override Guid ComponentGuid
        {
            get { return new Guid("6c5d1f8a-4932-4217-bbd3-6db0b5346061"); }
        }

        private class BuildingData
        {
            public List<BlockItem> PodiumBlocks { get; set; } = new List<BlockItem>();
            public Dictionary<string, List<BlockItem>> TowerBlocks { get; set; } = new Dictionary<string, List<BlockItem>>();
        }

        private class BlockItem
        {
            public RhinoObject Obj { get; set; }
            public string TowerId { get; set; }
            public double Z { get; set; }
        }
    }
}
